File: src/client/java/com/somefrills/features/misc/glowblock/GlowBlock.java
package com.somefrills.features.misc.glowblock;

import com.somefrills.config.FrillsConfig;
import com.somefrills.events.ServerJoinEvent;
import com.somefrills.events.WorldRenderEvent;
import com.somefrills.features.core.Feature;
import com.somefrills.misc.RenderColor;
import meteordevelopment.orbit.EventHandler;
import net.minecraft.block.Block;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Box;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;

import static com.somefrills.Main.mc;

public class GlowBlock extends Feature {
    private BlockScanner blockScanner;
    private final List<Block> targetBlocks = new ArrayList<>();
    public GlowBlock() {
        super(FrillsConfig.instance.misc.glowBlock.enabled);
    }

    @Override
    public void onEnable() {
        if(blockScanner == null) {
            this.blockScanner = new BlockScanner();
        }
        blockScanner.reset();
    }

    @Override
    public void onDisable() {
        if(blockScanner != null) {
            blockScanner.clearResultsOnly();
        }
    }

    @EventHandler
    public void onServerJoin(ServerJoinEvent event) {
        setEnabled(false);
    }

    public void addBlock(Block block) {
        if (targetBlocks.contains(block)) return;
        targetBlocks.add(block);

        if (blockScanner != null) {
            blockScanner.reset();
        }
    }

    public void removeBlock(Block block) {
        targetBlocks.remove(block);

        if (blockScanner != null) {
            blockScanner.removeBlockFromResults(block);
        }
    }

    public void clear() {
        targetBlocks.clear();

        if (blockScanner != null) {
            blockScanner.clearResultsOnly();
        }
    }

    public List<Block> getTargetBlocks() {
        return targetBlocks;
    }

    @EventHandler
    public void onWorldRender(WorldRenderEvent event) {
        if (mc.world == null || event.camera == null) return;
        if (targetBlocks.isEmpty()) return;

        Set<BlockPos> glowingBlocks = blockScanner.scanRenderedChunks(targetBlocks);
        for (BlockPos pos : glowingBlocks) {
            Box box = new Box(pos);
            event.drawOutline(box, true, new RenderColor(255, 255, 0, 128));
        }
    }

}

File: src/main/kotlin/com/somefrills/features/misc/glowblock/GlowBlock.kt
package com.somefrills.features.misc.glowblock

import com.somefrills.Main.mc
import com.somefrills.config.FrillsMod

import com.somefrills.events.ServerJoinEvent
import com.somefrills.events.WorldRenderEvent
import com.somefrills.features.core.Feature
import com.somefrills.modules.FrillsFeature
import com.somefrills.misc.RenderColor
import com.somefrills.events.core.EventHandle
import net.minecraft.world.level.block.Block
import net.minecraft.world.phys.AABB

@FrillsFeature
object GlowBlock : Feature(FrillsMod.config.misc.glowBlock.enabled) {
    private var blockScanner: BlockScanner? = null

    @JvmField
    val targetBlocks: MutableList<Block> = ArrayList<Block>()

    public override fun onActivate() {
        if (blockScanner == null) {
            this.blockScanner = BlockScanner()
        }
        blockScanner?.reset()
    }

    public override fun onDeactivate() {
        blockScanner?.clearResultsOnly()
    }

    @EventHandle
    fun onServerJoin(event: ServerJoinEvent) {
        enabled = false
    }

    fun addBlock(block: Block) {
        if (targetBlocks.contains(block)) return
        targetBlocks.add(block)

        if (blockScanner != null) {
            blockScanner!!.clearChunksOnly()
        }
    }

    fun removeBlock(block: Block) {
        targetBlocks.remove(block)

        if (blockScanner != null) {
            blockScanner!!.removeBlockFromResults(block)
        }
    }

    fun clear() {
        targetBlocks.clear()
        blockScanner?.clearResultsOnly()
    }

    @EventHandle
    fun onWorldRender(event: WorldRenderEvent) {
        if (mc.level == null) return
        if (targetBlocks.isEmpty()) return

        val glowingBlocks = blockScanner?.scanRenderedChunks(targetBlocks) ?: return
        for (pos in glowingBlocks) {
            val box = AABB(pos)
            event.drawOutline(box, true, RenderColor(255, 255, 0, 128))
        }
    }
}
